### PR TITLE
Add testEmbeddedObjectQuery

### PR DESCRIPTION
**Pull request by ericjordanmossman**
I believe there are no tests for queries with embedded object properties in the predicate ([RCOCOA-819](https://jira.mongodb.org/browse/RCOCOA-819)). This PR tries to address that.

I could have exhaustively tested every permutation of a predicate, but with embedded objects. I didn't do this because adding embedded object cases to every existing test would muddy their unit scope. And Recreating existing tests with embedded objects would make QueryTests.m less readable and maintainable.

So I chose to add this single test which hits on 1:1 relationship, embedded objects array, compound queries, aggregate operator, set operator, and sorting. The logic being, **_If_** this test demonstrates embedded objects largely function like normal objects, and every normal query is exhaustively tested (all datatypes, query combos) _then_ that would also apply to embedded objects.

But this means that the test isn't completely integrated. For example: 
This test should mostly catch breaking changes to the `RLMEmbeddedObject` class. 
We have tests to catch if a `CONTAINS` queries breaks on non-embedded object string properties.
But we _wouldn't_ have an integrated test that catches when `CONTAINS` breaks on string data types for embedded objects _only_.

Would something like this PR be sufficient for now? Or should I try to cover more cases?

**Some notes**

- I didn't make changes to any swift files because there are no swift query tests. I didn't think there was a need given queries are based on objc (NSPredicate).

- The object server tests that fail also fail in v10 branch.

- I used `EmbeddedIntParentObject` and `EmbeddedIntObject` from `RLMTestObjects.h` because that's what's used in the KVO tests.

### CODE REVIEW

File: Realm/Tests/QueryTests.m [hunk 1]
```diff
@@ -565,6 +565,36 @@ - (void)verifySort:(RLMRealm *)realm column:(NSString *)column ascending:(BOOL)a
     XCTAssertEqualObjects(obj[column], val, @"%@", column);
 }
 
+- (void)testEmbeddedObjectQuery
+{
```
Comment (jsflax) [inline Realm/Tests/QueryTests.m:569]: This brace should always be on the same line that the method is declared.

File: Realm/Tests/QueryTests.m [hunk 2]
```diff
@@ -565,6 +565,36 @@ - (void)verifySort:(RLMRealm *)realm column:(NSString *)column ascending:(BOOL)a
...
+    EmbeddedIntParentObject *obj2 = [EmbeddedIntParentObject createInRealm:realm withValue:@[@7, @[@8], @[@[@9]]]];
+    [realm commitWriteTransaction];
+    
+    NSArray *(^asArray)(RLMResults *) = ^(RLMResults *results) {
+        return [[self evaluate:results] valueForKeyPath:@"self"];
+    };
+    
+    // Query parent objects based on property of embedded object
+    RLMResults *r0 = [EmbeddedIntParentObject objectsWhere:@"object.intCol = 2"];
+    XCTAssertEqualObjects(asArray(r0), (@[ obj0 ]));
```
Comment (jsflax) [inline Realm/Tests/QueryTests.m:583]: The as array lambda you have there is a bit awkward and there is simpler way to achieve this:
```suggestion
    XCTAssertEqualObjects(r0[0], obj0);
```
You can do this with all of the proceeding equality checks as well.

File: Realm/Tests/QueryTests.m [hunk 3]
```diff
@@ -565,6 +565,36 @@ - (void)verifySort:(RLMRealm *)realm column:(NSString *)column ascending:(BOOL)a
...
+    // Query parent objects based on array of embedded objects
+    RLMResults *r1 = [EmbeddedIntParentObject objectsWhere:@"ANY array.intCol > 4"];
+    XCTAssertEqualObjects(asArray(r1), (@[ obj1, obj2 ]));
+
+    // Compound query using two different embedded object properties
+    RLMResults *r2 = [EmbeddedIntParentObject objectsWhere:@"ANY array.intCol > 4 and object.intCol = 5"];
+    XCTAssertEqualObjects(asArray(r2), (@[ obj1 ]));
+
+    // Aggregate query on embedded object array, sort using embedded object key path
+    RLMResults *r3 = [[EmbeddedIntParentObject objectsWhere:@"array.@max.intCol < 9"] sortedResultsUsingKeyPath:@"object.intCol" ascending:NO];
```
Comment (jsflax) [inline Realm/Tests/QueryTests.m:594]: Nitpick, but this line is rather long– you can probably add a newline in front of `sortedResultsUsingKeyPath`.

**Comment by jsflax** [comment]
This is definitely a good start– I would look to the `testQueryWithObjects` test and the `testMultiLevel...` series of tests for more ideas on cases.

The most robust amount of testing would ensure that the EmbeddedObjects can be queried as embedded objects, and that they also function the same as Objects w.r.t queries. There are probably ways to abstract out the existing object tests for the sake of DRYness.

Also, be sure to merge the latest v10. All tests on v10 are passing at this point.

**Comment by tgoyne** [comment]
You could try doing something like AsyncQueryTests or NullQueryTests to run the full suite of tests on an embedded object type, but I'm not sure there's a lot of value to doing so. With the current design and implementation of embedded objects it'd be very hard to have some types of queries work but others not. We mostly just need tests that verify that queries on embedded object types work at all, and tests which go significantly beyond that would be more of future-proofing in case we redesign things later to have the implementations diverge more.

**Comment by ericjordanmossman** [comment]
I've created a ticket to revisit exhaustive testing later.
I'm moving this PR out of draft to at least get a preliminary test active and to clean out the ongoing PR page.

**Failing tests**
[Jenkins: osx-object-server Release 11.3](https://ci.realm.io/job/objc_pr/configuration=Release,target=osx-object-server,xcode_version=11.3/7186/consoleFull) (Known Failure):
<details>

> Oct 07 13:06:56 Test session results, code coverage, and logs:
> Oct 07 13:06:56 	/Users/realm/workspace/objc_pr/configuration/Release/target/osx-object-server/xcode_version/11.3/build/DerivedData/Realm/Logs/Test/Test-Object Server Tests-2020.10.07_13-00-28--0700.xcresult
> Oct 07 13:06:56 
> Oct 07 13:06:56 Failing tests:
> Oct 07 13:06:56 	ObjectServerTests:
> Oct 07 13:06:56 		-[RLMObjectServerTests testMultipleRegisterDevice]
> Oct 07 13:06:56 
> Oct 07 13:06:56 ** TEST FAILED **

</details>

[Jenkins: osx-object-server Release 12.0](https://ci.realm.io/job/objc_pr/configuration=Release,target=osx-object-server,xcode_version=12.0/7186/consoleFull) (Same):
<details>

> Oct 07 13:05:57 Test session results, code coverage, and logs:
> Oct 07 13:05:57 	/Users/realm/workspace/objc_pr/configuration/Release/target/osx-object-server/xcode_version/12.0/build/DerivedData/Realm/Logs/Test/Test-Object Server Tests-2020.10.07_12-59-35--0700.xcresult
> Oct 07 13:05:57 
> Oct 07 13:05:57 Failing tests:
> Oct 07 13:05:57 	ObjectServerTests:
> Oct 07 13:05:57 		-[RLMObjectServerTests testMultipleRegisterDevice]
> Oct 07 13:05:57 
> Oct 07 13:05:57 ** TEST FAILED **

</details>

**Review by tgoyne — APPROVED** [review]

**Pull request merged**